src/adapters/pnFetch.ts

```diff
@@ -0,0 +1,280 @@
+/* eslint-disable max-params */
+import { URL } from "url";
+import { flow, pipe } from "fp-ts/lib/function";
+import * as t from "io-ts";
+import * as E from "fp-ts/Either";
+import * as TE from "fp-ts/TaskEither";
+import * as RA from "fp-ts/ReadonlyArray";
+import * as O from "fp-ts/Option";
+import { NonEmptyString } from "@pagopa/ts-commons/lib/strings";
+import { ProblemJson } from "@pagopa/ts-commons/lib/responses";
+import { Response as NodeResponse } from "node-fetch";
+import { NotificationDocument } from "generated/piattaforma-notifiche/NotificationDocument";
+import { NotificationAttachmentDownloadMetadataResponse } from "generated/piattaforma-notifiche/NotificationAttachmentDownloadMetadataResponse";
+import { match } from "ts-pattern";
+import { FullReceivedNotification } from "../../generated/piattaforma-notifiche/FullReceivedNotification";
+import { ThirdPartyAttachment } from "../../generated/third-party-service/ThirdPartyAttachment";
+import { ThirdPartyMessageDetail } from "../../generated/third-party-service/ThirdPartyMessageDetail";
+import { PnAPIClient } from "../clients/pn-client";
+import { errorsToError } from "../utils/errorsFormatter";
+import { pathParamsFromUrl } from "../types/pathParams";
+
+const getPath = (input: RequestInfo): string =>
+  pipe(typeof input === "string" ? input : input.url, i => new URL(i).pathname);
+
+export const ThirdPartyMessagesUrl = pathParamsFromUrl(
+  RegExp("^[/]+messages[/]+([^/]+)$"),
+  ([id]) => `/messages/${id}`
+);
+
+export const ThirdPartyAttachmentUrl = pathParamsFromUrl(
+  RegExp("^[/]+messages[/]+([^/]+)/(.+)$"),
+  ([id, url]) => `/messages/${id}/${url}`
+);
+
+export const PnDocumentUrl = pathParamsFromUrl(
+  RegExp(
+    "[/]+delivery[/]+notifications[/]+sent[/]+([^/]+)[/]+attachments[/]+documents[/]+([^/]+)$"
+  ),
+  ([iun, docIdx]) =>
+    `/delivery/notifications/sent/${iun}/attachments/documents/${docIdx}`
+);
+
+const WithFiscalCode = t.interface({ fiscal_code: NonEmptyString });
+type WithFiscalCode = t.TypeOf<typeof WithFiscalCode>;
+
+const retrieveNotificationDetails = (
+  origFetch: typeof fetch,
+  pnUrl: string,
+  pnApiKey: string,
+  headers: WithFiscalCode,
```

Comment:
why not passing directly `fiscalCode: FiscalCode` here?

Reply:
done. thank you!